File: src/content/docs/topics/cs/coding-problems/math-and-geometry/073-set-matrix-zeroes.go
package main

import (
	"fmt"
	"reflect"
)

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func setZeroes(matrix [][]int) {
	// TODO: implement
}

func runTests() {
	m := [][]int{{1, 1, 1}, {1, 0, 1}, {1, 1, 1}}
	setZeroes(m)
	assert(reflect.DeepEqual(m, [][]int{{1, 0, 1}, {0, 0, 0}, {1, 0, 1}}))
	m2 := [][]int{{0, 1, 2, 0}, {3, 4, 5, 2}, {1, 3, 1, 5}}
	setZeroes(m2)
	assert(reflect.DeepEqual(m2, [][]int{{0, 0, 0, 0}, {0, 4, 5, 0}, {0, 3, 1, 0}}))
	m3 := [][]int{{1}}
	setZeroes(m3)
	assert(reflect.DeepEqual(m3, [][]int{{1}}))
	m4 := [][]int{{0}}
	setZeroes(m4)
	assert(reflect.DeepEqual(m4, [][]int{{0}}))
	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/math-and-geometry/073-set-matrix-zeroes.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function setZeroes(matrix: number[][]): void {
    // TODO: implement (in place)
}

function _runTests(): void {
    const m = [[1, 1, 1], [1, 0, 1], [1, 1, 1]];
    setZeroes(m);
    assert(JSON.stringify(m) === JSON.stringify([[1, 0, 1], [0, 0, 0], [1, 0, 1]]));

    const m2 = [[0, 1, 2, 0], [3, 4, 5, 2], [1, 3, 1, 5]];
    setZeroes(m2);
    assert(JSON.stringify(m2) === JSON.stringify([[0, 0, 0, 0], [0, 4, 5, 0], [0, 3, 1, 0]]));

    const m3 = [[1]];
    setZeroes(m3);
    assert(JSON.stringify(m3) === JSON.stringify([[1]]));

    const m4 = [[0]];
    setZeroes(m4);
    assert(JSON.stringify(m4) === JSON.stringify([[0]]));

    const t0 = performance.now();
    const bigM: number[][] = Array.from({ length: 100 }, (_, i) =>
        Array.from({ length: 100 }, (_, j) => i * 100 + j)
    );
    bigM[50][50] = 0;
    setZeroes(bigM);
    console.log(`perf setZeroes 100x100 matrix: ${(performance.now() - t0).toFixed(1)}ms`);

    console.log('all tests pass');
}

_runTests();
